Serverless internal docker registry node port - docs (#17470)
kyma-project/kyma: docs/05-technical-reference/00-architecture/svls-02-internal-registry.md

---
title: Internal Docker Registry
---

> Internal docker registry is used for serverless to push built function images which can be used to deploy function without using 3rd party service.

Reviewer: I'm missing an opening like:
```
Kyma serverless module by default comes with internal docker registry deployment that is used to store function container images w/o necessity to  use 3rd party registry

Internal docker  is not recommended for production, as

is not deployed in HA setup
has limited storage space and no garbage collection of orphaned images
Still, it is very convenient for development and getting first time experience of kyma serverless.
```

Reviewer: please provide proper punctuation

The following diagram illustrates how it works.

![Serverless architecture](./assets/svls-internal-registry.svg)

1. Build job pushes function image to docker registry using in-cluster URL
2. Internal docker registry URL is resolved by K8s DNS to real ip address

> 3. Kubelet fetch image using URL: `localhost:{node_port}/{image}`

Reviewer: I think we should mention also the fact that when multiple nodes are used, the nodePort routing will assure that the request will reach the node where internal registry is deployed

The internal Docker registry is not recommended for production, as it's not deployed in the High Availability (HA) setup and has limited storage space and no garbage collection of the orphaned images.

Still, it is very convenient for development and getting first-time experience with Kyma Serverless.

See the following diagram to learn how it works:

![Serverless architecture](./assets/svls-internal-registry.svg)

1. Build job pushes the Function image to the Docker registry using the in-cluster URL.
2. The Kubernetes DNS resolves the internal Docker registry URL to the actual IP address.
3. [kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) fetches the image using the URL: `localhost:{node_port}/{image}`.
4. NodePort allows kubelet to get into the cluster network and translate `localhost` to `internal-registry.kyma-system.svc.cluster.local` and ask the Kubernetes DNS to resolve the name.
5. The Kubernetes DNS service resolves the name and provides the IP of the internal Docker registry.

> **NOTE:** kubelet cannot resolve the in-cluster URL. That's why Serverless uses the NodePort service.

Author: ```suggestion
**NOTE:** kubelet cannot resolve the in-cluster URL. That's why Serverless uses the NodePort service.

```

---
title: Internal Docker Registry
---

> Kyma serverless module by default comes with internal docker registry deployment that is used to store function container images w/o necessity to  use 3rd party registry

Author: ```suggestion
By default, the Kyma Serverless module comes with the internal Docker registry, which stores the Function container images without using the third-party registry.
```

Internal docker  is not recommended for production, as

is not deployed in HA setup
has limited storage space and no garbage collection of orphaned images
Still, it is very convenient for development and getting first time experience of kyma serverless.

The following diagram illustrates how it works.

![Serverless architecture](./assets/svls-internal-registry.svg)

1. Build job pushes function image to docker registry using in-cluster URL
2. Internal docker registry URL is resolved by K8s DNS to real ip address

> 3. Kubelet fetch image using URL: `localhost:{node_port}/{image}`

Reviewer 2: ```suggestion
3. [kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) fetches the image using the URL: `localhost:{node_port}/{image}`.
```

> 4. NodePort allows kubelet to get into the cluster network and translate `localhost` to `internal-registry.kyma-system.svc.cluster.local` and ask the k8s dns to resolve the name

Reviewer 2: ```suggestion
4. NodePort allows kubelet to get into the cluster network and translate `localhost` to `internal-registry.kyma-system.svc.cluster.local` and ask the Kubernetes DNS to resolve the name.
```

> 5. K8s DNS service resolve the name and provide the IP of `internal docker registry`

Reviewer 2: ```suggestion
5. The Kubernetes DNS service resolves the name and provides the IP of the internal Docker registry.
```
Do we need the `` ticks for the internal Docker registry?

Author: No

> **NOTE:** Kubelet cannot resolve in-cluster URL that's why serverless uses NodePort service.

Reviewer 2: ```suggestion
**NOTE:** kubelet cannot resolve the in-cluster URL. That's why Serverless uses the NodePort service.
```
Maybe we can move this note higher? Under point 3 or 4?

> **NOTE:** NodePort service routing assure that pull request reaches the internal docker registry regardless of whether it is from different node.

Reviewer 2: ```suggestion
**NOTE:** The NodePort service routing assures that the pull request reaches the internal Docker registry regardless of whether it is from a different node.
```
Maybe we can also move this note higher, where it suits better? Maybe under point 4?

Author: I don't have a strong opinion, but I would like to keep it at the bottom to not interfere with the explanation of how it works.
